crates/ra_hir/src/expr/validation.rs

```diff
@@ -87,4 +97,58 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
             })
         }
     }
+
+    fn validate_results_in_tail_expr(&mut self, id: ExprId, db: &impl HirDatabase) {
+        let mismatch = match self.infer.type_mismatch_for_expr(id) {
+            Some(m) => m,
+            None => return,
+        };
+
+        let std_result_path = Path {
+            kind: PathKind::Abs,
+            segments: vec![
+                PathSegment { name: name::STD, args_and_bindings: None },
+                PathSegment { name: name::RESULT_MOD, args_and_bindings: None },
+                PathSegment { name: name::RESULT_TYPE, args_and_bindings: None },
+            ],
+        };
+
+        let resolver = self.func.resolver(db);
+        let std_result_enum =
+            match resolver.resolve_path_segments(db, &std_result_path).into_fully_resolved() {
+                PerNs { types: Some(Resolution::Def(ModuleDef::Enum(e))), .. } => e,
+                _ => return,
+            };
+
+        let std_result_type = std_result_enum.ty(db);
+
+        fn enum_from_type(ty: &Ty) -> Option<Enum> {
+            match ty {
+                Ty::Apply(t) => match t.ctor {
+                    TypeCtor::Adt(AdtDef::Enum(e)) => Some(e),
+                    _ => None,
+                },
+                _ => None,
+            }
+        }
+
+        if enum_from_type(&mismatch.expected) != enum_from_type(&std_result_type) {
```

Comment:
I think this is slightly roundabout: `enum_from_type(&std_result_type)` returns exactly `std_result_enum` which we have started from! 

Let's simplify to 

```
let ret =  match mismatch.expected {
    Ty::Apply(ret @ Apply { ctor, ..}) if ctor == std_result_enum => ret,
    _ => return,
}
```